### ISSUE TITLE
Windows: crowdsale participation doesn't yield expected values

### ISSUE THREAD

**Issue by dexX7** [OmniLayer/omnicore#124]
- start QT `-regtest` mode.
- go to debug console
- `getnewaddress` -> `mqGdJobtXrrrLn4A8tsTbdb9sCYncgxbut`
- `getnewaddress` -> `mnqvUM848To2vKjkni4AG8PWy9VaFN2BtF`
- `setgenerate true 101`
- send 3 BTC to `mqGdJobtXrrrLn4A8tsTbdb9sCYncgxbut`
- use coin control to select the "from" addresses for the following BTC sends:
- from `mqGdJobtXrrrLn4A8tsTbdb9sCYncgxbut` send 1 BTC to `moneyqMan7uh8FqdCA2BV5yZ8qVrc9ikLP`, set `mqGdJobtXrrrLn4A8tsTbdb9sCYncgxbut` as custom change address
- from `mqGdJobtXrrrLn4A8tsTbdb9sCYncgxbut` send 1 BTC to `mnqvUM848To2vKjkni4AG8PWy9VaFN2BtF`, set `mqGdJobtXrrrLn4A8tsTbdb9sCYncgxbut` as custom change address
- `setgenerate true 1`
- restart the client (note: Exodus purchase doesn't refresh UI)
- `omni_sendrawtx mnqvUM848To2vKjkni4AG8PWy9VaFN2BtF 000000330100020000000000004d446976000000000000017fffffffffffffff00000001ccd403f00000`
- `setgenerate true 1` (note: crowdsale creation doesn't show up in UI)
- `omni_send mqGdJobtXrrrLn4A8tsTbdb9sCYncgxbut mnqvUM848To2vKjkni4AG8PWy9VaFN2BtF 1 0.1`
- `setgenerate true 1` (note: this updates the UI finally)
- `omni_gettransaction [txid of the send]`

The result I see:
```java
{
  "txid": "4a5f6f4336172a25b23c0ce50f4f2b2a51e3585672d2b2984961c64cf187b2e3",
  "sendingaddress": "mqGdJobtXrrrLn4A8tsTbdb9sCYncgxbut",
  "referenceaddress": "mnqvUM848To2vKjkni4AG8PWy9VaFN2BtF",
  "ismine": true,
  "confirmations": 1,
  "fee": "0.00002570",
  "blocktime": 1436413923,
  "valid": true,
  "version": 0,
  "type_int": 0,
  "type": "Crowdsale Purchase",
  "propertyid": 1,
  "divisible": true,
  "amount": "0.10000000",
  "purchasedpropertyid": 3,
  "purchasedpropertyname": "MDiv",
  "purchasedpropertydivisible": true,
  "purchasedtokens": "9223372038.98877165", <--- should be: 9223372036.85477580
  "issuertokens": "0.00000000"
}
```
And for the context `omni_getcrowdsale 3 true`:
```java
{
  "propertyid": 3,
  "name": "MDiv",
  "active": true,
  "issuer": "mnqvUM848To2vKjkni4AG8PWy9VaFN2BtF",
  "propertyiddesired": 1,
  "tokensperunit": "92233720368.54775807",
  "earlybonus": 0,
  "percenttoissuer": 0,
  "starttime": 1436413795,
  "deadline": 7731414000,
  "amountraised": "0.00000000",
  "tokensissued": "9223372038.98877165", <--- should be: 9223372036.85477580
  "addedissuertokens": "0.00000000",
  "participanttransactions": [
  {
    "txid": "4a5f6f4336172a25b23c0ce50f4f2b2a51e3585672d2b2984961c64cf187b2e3",
    "amountsent": "0.10000000",
    "participanttokens": "9223372038.98877165", <--- should be: 9223372036.85477580
    "issuertokens": "0.00000000"
  }]
}
```

Investing different amounts may yield wrong values, too:
```
| Invested   | Expected             | Got                  | Correct? |
| ---------- | -------------------- | -------------------- | -------- |
| 0.00000001 | 922.33720368         | 922.33720388         |    NO    |
| 0.0000001  | 9223.37203685        | 9223.37203897        |    NO    |
| 0.000001   | 92233.72036854       | 92233.72038987       |    NO    |
| 0.00001    | 922337.20368547      | 922337.20389886      |    NO    |
| 0.0001     | 9223372.03685477     | 9223372.03898876     |    NO    |
| 0.001      | 92233720.36854775    | 92233720.38988770    |    NO    |
| 0.01       | 922337203.68547758   | 922337203.89887715   |    NO    |
| 0.1        | 9223372036.85477580  | 9223372038.98877165  |    NO    |
                                               ^^^^^^^^^^
| 1.0        | 92233720368.54775807 | 92233720368.54775807 |   YES    |
```

**Comment by zathras-crypto** [comment]
Can I just clarify, is this an issue at the RPC layer or did the core logic actually grant an incorrect amount of tokens (what does log show?)

**Comment by dexX7** [comment]
Unfortunally it's not an RPC issue, but the numbers also don't match in the log.

Interestingly, when using the 32 bit build, it works, as I found out a few seconds ago.

This seems to be limited to Win x64.

I have a build with more debug output in a few minutes ready.

**Comment by zathras-crypto** [comment]
> Unfortunally it's not an RPC issue, but the numbers also don't match in the log.

Yikes - how is it possible that Travis CI is passing everything if Winx64 builds are giving bad data?

**Comment by zathras-crypto** [comment]
GREAT catch btw :)

**Comment by dexX7** [comment]
We run the OmniJ tests only during the Linux x64 build, because it's not (yet) possible to run .exe files.

But this was actually discovered by running the OmniJ tests locally on Windows. :)

(download OmniJ -> `gradlew.bat :omnij-rpc:regTest`)

**Comment by zathras-crypto** [comment]
I'd be very interested in what the 0.0.9.1 Windows binary showed for that test also

**Comment by dexX7** [comment]
Hmmm...

Good (Ubuntu x64):
```
	         version: 0, class C
	            type: 0 (Simple Send)
	        property: 1 (MSC)
	           value: 0.00000001
update_tally_map(mvGHJhZNh8icadicuueTkK6JSmLszSqDiU, 1=0x1, -1, ttype=0): before=1, after=0
update_tally_map(mhjx2VTHfofKVMFbmJUHWz848dV4udHd6a, 1=0x1, +1, ttype=0): before=0, after=1
INVESTMENT SEND to Crowdsale Issuer: mhjx2VTHfofKVMFbmJUHWz848dV4udHd6a
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
calculateFundraiser(): bonusSeconds_ = 6294998840
calculateFundraiser(): weeks_ = 10408397552910052
calculateFundraiser(): ebPercentage_ = 0
calculateFundraiser(): bonusPercentage_ = 1000000000000
calculateFundraiser(): issuerPercentage_ = 0
calculateFundraiser(): createdTokens = 9223372036854775807000000000000
calculateFundraiser(): issuerTokens = 0
calculateFundraiser(): createdTokens_int = 92233720368
calculateFundraiser(): issuerTokens_int = 0
calculateFundraiser(): newTotalCreated = 92233720368
calculateFundraiser(): createdTokens_int -> 92233720368
calculateFundraiser(): issuerTokens_int -> 0
logicMath_SimpleSend(): txdata[] = {1, 1436415160, 92233720368, 0}
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
```

Bad (Win x64):
```
	         version: 0, class C
	            type: 0 (Simple Send)
	        property: 1 (MSC)
	           value: 0.00000001
update_tally_map(moohyeXkGP917n5NgLomTsosA3eYCGFMct, 1=0x1, -1, ttype=0): before=1, after=0
update_tally_map(mqgvgskaAn1W4Ed2srBgFcQPsrmNT68waV, 1=0x1, +1, ttype=0): before=0, after=1
INVESTMENT SEND to Crowdsale Issuer: mqgvgskaAn1W4Ed2srBgFcQPsrmNT68waV
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
calculateFundraiser(): bonusSeconds_ = 6294996921
calculateFundraiser(): weeks_ = 10408394379960317
calculateFundraiser(): ebPercentage_ = 0
calculateFundraiser(): bonusPercentage_ = 1000000000000
calculateFundraiser(): issuerPercentage_ = 0
calculateFundraiser(): createdTokens = 9223372036854775807000000000000
calculateFundraiser(): issuerTokens = 0
calculateFundraiser(): createdTokens_int = 92233720388 <-------
calculateFundraiser(): issuerTokens_int = 0
calculateFundraiser(): newTotalCreated = 92233720388
calculateFundraiser(): createdTokens_int -> 92233720388
calculateFundraiser(): issuerTokens_int -> 0
logicMath_SimpleSend(): txdata[] = {1, 1436417079, 92233720388, 0}
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
```

That's the line appearingly causing it: [src/omnicore/omnicore.cpp#L494](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/omnicore/omnicore.cpp#L494):

```c++
cpp_int createdTokens_int = createdTokens / (precision_ * satoshi_precision_);
```

With numbers:
```c++
cpp_int createdTokens_int = 9223372036854775807000000000000 / (precision_ * satoshi_precision_);
```

Unfortunally I didn't log `precision_` and `satoshi_precision_`, but assuming these are correct:

```c++
cpp_int createdTokens_int = 9223372036854775807000000000000 / (1000000000000 * 100000000)
                          = 9223372036854775807000000000000 / 100000000000000000000
                          = 92233720368
```

**Comment by zathras-crypto** [comment]
Nice debugging dude, ```cpp_int``` is boost multiprecision so I would have thought we shouldn't see any difference between Windows and Linux there?

**Comment by dexX7** [comment]
D'oh, the log output isn't really useful though. I logged the other values, too, but there is nothing unusual, expect the actual result:

```
calculateFundraiser(): precision_ = 1000000000000
calculateFundraiser(): satoshi_precision_ = 100000000
calculateFundraiser(): createdTokens = 9223372036854775807000000000000
calculateFundraiser(): (precision_ * satoshi_precision_) = 100000000000000000000
calculateFundraiser(): createdTokens_int = 92233720388 <--- no!
```

**Comment by zathras-crypto** [comment]
So, just to make sure on same page here - it seems that we are saying that boost::multiprecision gives different answers for ```9223372036854775807000000000000 / 100000000000000000000``` on Windows vs Linux?  If so that is a big "oh no" moment because it pretty much rules out the usage of that library :(

**Comment by dexX7** [comment]
Yes, this seems to be the case, even though it shouldn't.

http://www.boost.org/doc/libs/1_58_0/libs/multiprecision/doc/html/boost_multiprecision/tut/ints/cpp_int.html

There is a note:

> Since the precision of cpp_int_backend is necessarily limited when the allocator parameter is void, care should be taken to avoid numeric overflow when using this type unless you actually want modulo-arithmetic behavior.

Does this apply? Could this somehow be related? I'm going to replace the `cpp_int` with fixed integers now, but it really takes a long time until the results are ready, which isn't very helpful right now.

**Comment by zathras-crypto** [comment]
> Does this apply? Could this somehow be related?

Honest answer, I don't know 

> I'm going to replace the cpp_int with fixed integers now

Thanks heaps dude, I can't see anywhere else it's used outside of crowdsale calculation, can you?

**Comment by dexX7** [comment]
Hm.. making 100 % sure that the same types etc. are used didn't do the trick either:

```c++
cpp_int createdTokens_int = cpp_int(createdTokens) / (cpp_int(precision_) * cpp_int(satoshi_precision_));
int256_t createdTokens_256 = int256_t(createdTokens) / (int256_t(precision_) * int256_t(satoshi_precision_));
```
Both results are still wrong.

**Comment by zathras-crypto** [comment]
> Both results are still wrong.

Are they the same results?  (ie does ```int256_t``` give same result as ```cpp_int```)?

**Comment by dexX7** [comment]
Yes, they create the same results. It comes basically down to this:
```
1)         9223372036854775807000000000000 / 100000000000000000000 = 92233720388         FAIL
2)        92233720368547758070000000000000 / 100000000000000000000 = 922337203897        FAIL
3)       922337203685477580700000000000000 / 100000000000000000000 = 9223372038987       FAIL
4)      9223372036854775807000000000000000 / 100000000000000000000 = 92233720389886      FAIL
5)     92233720368547758070000000000000000 / 100000000000000000000 = 922337203898876     FAIL
6)    922337203685477580700000000000000000 / 100000000000000000000 = 9223372038988770    FAIL
7)   9223372036854775807000000000000000000 / 100000000000000000000 = 92233720389887715   FAIL
8)  92233720368547758070000000000000000000 / 100000000000000000000 = 922337203898877165  FAIL
9) 922337203685477580700000000000000000000 / 100000000000000000000 = 9223372036854775807  OK
```

Right now we know:

- this bug is only present in Windows
- this bug is only present in the 64 bit build, but not in the 32 bit build
- it doesn't seem to be related to the length of the number, given that check 9) looks good
- `cpp_int`, `int256_t` or a mix of `cpp_int` and `int128_t` yield the same results

What really makes me wonder.. I would assume, and let's ignore that 1) - 8) are off, that each number over the other is `floor (num / 10)` (e.g. `floor (105 / 10) = 10`), but this isn't the case here:

```
922337203898877165
92233720389887715
```
... I would have expected `..715`, not `..716` for example.

**Comment by dexX7** [comment]
I switched to my Ubuntu VM and build there for Win x64:
```bash
$ git clone https://github.com/OmniLayer/omnicore.git
$ cd omnicore/
$ HOST=x86_64-w64-mingw32
$ make $MAKEJOBS -C depends HOST=$HOST
$ ./autogen.sh
$ ./configure --prefix=$PWD/depends/$HOST
$ make HOST=$HOST -j3
```

Here is the catch:

It works, and the results are valid. -_-"

Since the dependencies are the same (i.e. Boost 1.55 etc.), I guess it can be narrowed down to:

- other build configurations
- Debian vs. Ubuntu (although Gitian actually uses Ubuntu 12.04 in it's own VM afaik)

Since you built with Debian 8.4, and I with 9.1, and we ended up with similar results, trying a different version is probably pointless.

I'm running out of ideas.

**Comment by zathras-crypto** [comment]
> It works, and the results are valid. -_-"

OK, now I'm super confused...

> Since the dependencies are the same (i.e. Boost 1.55 etc.)

This is what I don't understand - how can the way ```cpp_int``` does math in Boost change based on anything other than the boost code?  If the deps are the same, I'm puzzled...

> Since you built with Debian 8.4, and I with 9.1

Assume you mean 7.4 & 8.1, and yeah I agree I don't think Debian version has much to do with it.

> other build configurations

Could this be compiler options or something changing how Boost is compiled and in turn how it functions?

**Comment by dexX7** [comment]
> Assume you mean ...

Ah, yup. :)

This is the file Gitian uses for the build process: [contrib/gitian-descriptors/gitian-win.yml](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/contrib/gitian-descriptors/gitian-win.yml)

I assume it would be possible to follow it step by step manually, although the environment probably plays a role as well.

> This is what I don't understand - how can the way cpp_int does math in Boost change based on anything other than the boost code?

Let's say it this way: I would expect it, too, especially given that Boost is often used, and Win x64 isn't rare either. I don't rule out it's somehow an issue we caused, but I also don't find any indication that this is the case.

Windows is a bit special though, because in contrast to Unix, `long` is [only 32 bit wide](https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models).

My first guess was that we lost precision somewhere, for example with:
```c++
int128_t precision_ = 1000000000000L;
```
... which should probably be:
```c++
int128_t precision_ = 1000000000000LL;
```
But I think the compiler takes care of it already, and the log output confirmed correct values.

And this doesn't explain why the Gitian build fails, and the local build works.

**Comment by dexX7** [comment]
Solved via #127. (for now...)
